File: src/praisonai-agents/praisonaiagents/agent/heartbeat.py
"""Heartbeat — runs an agent on a schedule, delivers results via callback.

Standalone wrapper class. Does NOT modify the Agent class.

Usage::

    from praisonaiagents import Agent, Heartbeat

    agent = Agent(instructions="Check server status")
    hb = Heartbeat(agent, schedule="every 30m", prompt="Report status")
    hb.start()   # blocking loop
    # hb.start(blocking=False)  # background thread
"""

import asyncio
import logging
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Optional, Union

logger = logging.getLogger(__name__)


@dataclass
class HeartbeatConfig:
    """Configuration for Heartbeat scheduler.

    Attributes:
        schedule: Human-friendly schedule expression. Supports:
            - Keywords: "hourly", "daily", "weekly"
            - Intervals: "every 30m", "every 6h", "*/30m", "*/10s"
            - Cron: "cron:0 7 * * *"
            - One-shot: "at:2026-03-01T09:00:00"
        prompt: Override prompt sent to agent each tick. None = use agent's default.
        on_result: Callback receiving (result_text: str). Default: log to stdout.
        on_error: "retry" | "skip" | callable(error). Default: "retry".
        max_retries: Max consecutive retries before skipping. Default: 3.
    """
    schedule: str = "hourly"
    prompt: Optional[str] = None
    on_result: Optional[Callable] = None
    on_error: Union[str, Callable] = "retry"
    max_retries: int = 3


class Heartbeat:
    """Standalone heartbeat coordinator.

    Runs an Agent on a schedule and delivers results via callback.
    Does NOT add any params to the Agent class.

    Usage::

        from praisonaiagents import Agent, Heartbeat

        agent = Agent(instructions="Monitor server health")
        hb = Heartbeat(agent, schedule="hourly")
        hb.start()
    """

    def __init__(
        self,
        agent,
        schedule: str = "hourly",
        prompt: Optional[str] = None,
        on_result: Optional[Callable] = None,
        on_error: Union[str, Callable] = "retry",
        max_retries: int = 3,
    ):
        self.agent = agent
        self.config = HeartbeatConfig(
            schedule=schedule,
            prompt=prompt,
            on_result=on_result,
            on_error=on_error,
            max_retries=max_retries,
        )
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._interval_seconds = self._resolve_interval(schedule)

    def start(self, blocking: bool = True) -> None:
        """Start the heartbeat loop.

        Args:
            blocking: If True, blocks the calling thread. If False,
                      runs in a background daemon thread.
        """
        self._running = True
        if blocking:
            self._loop()
        else:
            self._thread = threading.Thread(target=self._loop, daemon=True)
            self._thread.start()

    def stop(self) -> None:
        """Stop the heartbeat loop."""
        self._running = False

    @property
    def is_running(self) -> bool:
        return self._running

    # ── internals ────────────────────────────────────────────────────

    def _loop(self) -> None:
        """Main polling loop."""
        retries = 0
        while self._running:
            try:
                result = self._tick()
                retries = 0  # Reset on success
                self._deliver(result)
            except Exception as e:
                retries += 1
                logger.error(f"[heartbeat] {self.agent.name}: error on tick #{retries}: {e}")
                if callable(self.config.on_error):
                    self.config.on_error(e)
                elif self.config.on_error == "skip":
                    logger.warning(f"[heartbeat] Skipping error: {e}")
                else:  # "retry"
                    if retries >= self.config.max_retries:
                        logger.error(
                            f"[heartbeat] {self.agent.name}: max retries "
                            f"({self.config.max_retries}) reached, skipping."
                        )
                        retries = 0
            time.sleep(self._interval_seconds)

    def _tick(self) -> str:
        """Execute one heartbeat tick — run the agent and return result."""
        prompt = self.config.prompt or "Run your scheduled check."
        result = self.agent.start(prompt)
        return str(result) if result else ""

    def _deliver(self, result: str) -> None:
        """Deliver result via callback or log."""
        if self.config.on_result:
            self.config.on_result(result)
        else:
            logger.info(f"[heartbeat] {self.agent.name}: {result[:200]}")

    @staticmethod
    def _resolve_interval(schedule: str) -> float:
        """Convert schedule string to interval in seconds.

        Uses the scheduler parser if available, otherwise falls back to
        simple keyword parsing.
        """
        try:
            from ..scheduler.parser import parse_schedule
            sched = parse_schedule(schedule)
            if sched.every_seconds:
                return float(sched.every_seconds)
        except (ImportError, Exception):
            pass

        # Fallback keyword resolution
        keywords = {
            "hourly": 3600,
            "daily": 86400,
            "weekly": 604800,
        }
        lower = schedule.lower().strip()
        if lower in keywords:
            return float(keywords[lower])

        # Try parsing "every Xm/h/s" patterns
        import re
        match = re.match(r"(?:every\s+)?(\d+)\s*(s|m|h)", lower)
        if match:
            val, unit = int(match.group(1)), match.group(2)
            multipliers = {"s": 1, "m": 60, "h": 3600}
            return float(val * multipliers.get(unit, 60))

        return 3600.0  # Default to hourly

File: src/praisonai-ts/src/agent/heartbeat.ts
/**
 * Heartbeat — runs an agent on a schedule, delivers results via callback.
 *
 * Python parity: praisonaiagents/agent/heartbeat.py (`HeartbeatConfig`,
 * `Heartbeat`). Standalone wrapper; does NOT modify the Agent class.
 *
 * Python runs a thread that ticks, sleeps `interval` seconds, and repeats.
 * Here the loop is a chain of `setTimeout` calls that are `unref()`'d, so a
 * running heartbeat never keeps the Node process alive on its own, and
 * `stop()` clears the pending timer immediately.
 *
 * @example
 * const hb = new Heartbeat(agent, { schedule: 'every 30m', prompt: 'Report status' });
 * await hb.start();        // resolves when stop() is called
 * hb.start(false);         // background: returns after scheduling the first tick
 */

import { getLogger } from '../utils/logger';

const logger = getLogger('praisonaiagents.agent.heartbeat');

/** The slice of an agent the heartbeat needs: a name and `start(prompt)`. */
export interface HeartbeatAgent {
    name?: string;
    start(prompt: string): unknown;
}

/** Python `on_error`: "retry" | "skip" | callable(error). */
export type HeartbeatOnError = 'retry' | 'skip' | ((error: unknown) => void);

/** Options accepted by {@link HeartbeatConfig} and {@link Heartbeat}; the Python dataclass fields. */
export interface HeartbeatOptions {
    /**
     * Human-friendly schedule. Python: `schedule: str = "hourly"`. Supported:
     * keywords `hourly` | `daily` | `weekly`; intervals `every 30m`, `*\/30m`,
     * `30m`, `every 6h`, `*\/10s`, `2 hours`; raw seconds `"3600"`.
     * See {@link parseHeartbeatSchedule}.
     */
    schedule?: string;
    /** Prompt sent each tick. Python: `prompt: Optional[str] = None` (null = "Run your scheduled check."). */
    prompt?: string | null;
    /** Receives each tick's result text. Python: `on_result: Optional[Callable] = None` (null = log it). */
    onResult?: ((result: string) => void) | null;
    /** Error policy. Python: `on_error: Union[str, Callable] = "retry"`. */
    onError?: HeartbeatOnError;
    /** Consecutive failures before the retry counter resets. Python: `max_retries: int = 3`. */
    maxRetries?: number;
}

/**
 * Configuration for the Heartbeat scheduler.
 *
 * Python parity: `HeartbeatConfig` dataclass in agent/heartbeat.py, same
 * fields and defaults.
 */
export class HeartbeatConfig {
    readonly schedule: string;
    readonly prompt: string | null;
    readonly onResult: ((result: string) => void) | null;
    readonly onError: HeartbeatOnError;
    readonly maxRetries: number;

    constructor(config: HeartbeatOptions = {}) {
        this.schedule = config.schedule ?? 'hourly';
        this.prompt = config.prompt ?? null;
        this.onResult = config.onResult ?? null;
        this.onError = config.onError ?? 'retry';
        this.maxRetries = config.maxRetries ?? 3;
    }
}

const SCHEDULE_KEYWORDS: Record<string, number> = {
    hourly: 3600,
    daily: 86400,
    weekly: 604800,
};

// Same grammar as praisonaiagents/scheduler/parser.py `_INTERVAL_RE`, with the
// optional "every " prefix that heartbeat.py's fallback also accepts.
const INTERVAL_RE = /^(?:every\s+)?\*?\/?\*?(\d+)\s*(s|sec|second|seconds|m|min|minute|minutes|h|hr|hour|hours)$/i;

const UNIT_MULTIPLIER: Record<string, number> = {
    s: 1, sec: 1, second: 1, seconds: 1,
    m: 60, min: 60, minute: 60, minutes: 60,
    h: 3600, hr: 3600, hour: 3600, hours: 3600,
};

/** `setTimeout` cannot wait longer than this (2^31 - 1 ms, ~24.8 days). */
const MAX_TIMEOUT_MS = 2147483647;

/**
 * Convert a heartbeat schedule string to an interval in seconds.
 *
 * Python parity: `Heartbeat._resolve_interval` in agent/heartbeat.py, using
 * the interval grammar of scheduler/parser.py. Accepted forms:
 *   - keywords: `hourly`, `daily`, `weekly`
 *   - intervals: `every 30m`, `*\/30m`, `30m`, `every 6h`, `*\/10s`, `2 hours`
 *   - raw positive seconds: `3600`
 *
 * Deviation from Python, on purpose: Python silently falls back to hourly
 * for anything it cannot parse (including `cron:` and `at:` forms, which the
 * heartbeat cannot run). Here an unrecognised or non-positive schedule throws
 * so a misspelled schedule is caught at construction instead of running at
 * the wrong cadence.
 *
 * @throws {Error} when the schedule is empty, unrecognised, or resolves to zero seconds.
 */
export function parseHeartbeatSchedule(schedule: string): number {
    if (typeof schedule !== 'string' || !schedule.trim()) {
        throw new Error('Schedule expression cannot be empty');
    }
    const expr = schedule.trim();
    const lower = expr.toLowerCase();

    if (lower in SCHEDULE_KEYWORDS) {
        return SCHEDULE_KEYWORDS[lower];
    }

    let seconds: number | undefined;
    const match = INTERVAL_RE.exec(expr);
    if (match) {
        seconds = parseInt(match[1], 10) * UNIT_MULTIPLIER[match[2].toLowerCase()];
    } else if (/^\d+$/.test(expr)) {
        seconds = parseInt(expr, 10);
    }

    if (seconds === undefined) {
        throw new Error(
            `Unrecognised heartbeat schedule: ${JSON.stringify(schedule)}. ` +
            "Use 'hourly', 'daily', 'weekly', 'every 30m', '*/6h', '*/10s' or raw seconds. " +
            "Cron and one-shot ('cron:', 'at:') schedules are not intervals and cannot drive a heartbeat."
        );
    }
    if (!(seconds > 0)) {
        throw new Error(`Heartbeat schedule must be a positive interval, got ${JSON.stringify(schedule)}`);
    }
    return seconds;
}

/**
 * Standalone heartbeat coordinator: runs an agent on a schedule and delivers
 * results via callback. Does NOT add any params to the Agent class.
 *
 * Python parity: `Heartbeat(agent, schedule="hourly", prompt=None,
 * on_result=None, on_error="retry", max_retries=3)` in agent/heartbeat.py.
 * The keyword parameters travel in `options` (see {@link HeartbeatOptions}).
 */
export class Heartbeat {
    readonly agent: HeartbeatAgent;
    readonly config: HeartbeatConfig;

    private _running = false;
    private _timer: ReturnType<typeof setTimeout> | null = null;
    private _tickInFlight = false;
    private _retries = 0;
    private readonly _intervalSeconds: number;
    private _stopped: Promise<void> | null = null;
    private _resolveStopped: (() => void) | null = null;

    constructor(agent: HeartbeatAgent, options: HeartbeatOptions = {}) {
        this.agent = agent;
        this.config = new HeartbeatConfig(options);
        this._intervalSeconds = parseHeartbeatSchedule(this.config.schedule);
    }

    /** Interval between ticks in seconds, as resolved from `config.schedule`. */
    get intervalSeconds(): number {
        return this._intervalSeconds;
    }

    /** Python `is_running`. */
    get isRunning(): boolean {
        return this._running;
    }

    /**
     * Start the heartbeat loop. The first tick runs immediately, then every
     * interval, as in Python.
     *
     * @param blocking Python `start(blocking=True)`. When true the returned
     *   promise resolves only after {@link stop} is called and any in-flight
     *   tick has finished (the async counterpart of a blocking loop). When
     *   false the promise resolves as soon as the loop is scheduled, and the
     *   heartbeat keeps running in the background.
     */
    start(blocking: boolean = true): Promise<void> {
        if (!this._running) {
            this._running = true;
            this._retries = 0;
            this._stopped = new Promise<void>((resolve) => {
                this._resolveStopped = resolve;
            });
            void this._loop();
        }
        return blocking ? (this._stopped as Promise<void>) : Promise.resolve();
    }

    /** Stop the heartbeat loop; clears the pending timer so nothing fires again. */
    stop(): void {
        if (!this._running) {
            return;
        }
        this._running = false;
        if (this._timer !== null) {
            clearTimeout(this._timer);
            this._timer = null;
        }
        if (!this._tickInFlight) {
            this._settleStopped();
        }
    }

    // ── internals ────────────────────────────────────────────────────

    private _settleStopped(): void {
        const resolve = this._resolveStopped;
        this._resolveStopped = null;
        this._stopped = null;
        if (resolve) {
            resolve();
        }
    }

    /** One iteration of Python's `_loop`: tick, handle the outcome, sleep. */
    private async _loop(): Promise<void> {
        if (!this._running) {
            return;
        }
        this._tickInFlight = true;
        try {
            const result = await this._tick();
            this._retries = 0; // Reset on success
            this._deliver(result);
        } catch (error) {
            this._handleError(error);
        } finally {
            this._tickInFlight = false;
        }

        if (!this._running) {
            this._settleStopped();
            return;
        }
        const delayMs = Math.min(this._intervalSeconds * 1000, MAX_TIMEOUT_MS);
        const timer = setTimeout(() => {
            this._timer = null;
            void this._loop();
        }, delayMs);
        // Never keep the process alive just for the next tick.
        if (typeof (timer as any).unref === 'function') {
            (timer as any).unref();
        }
        this._timer = timer;
    }

    private _handleError(error: unknown): void {
        this._retries += 1;
        const name = this.agent.name ?? 'agent';
        void logger.error(`[heartbeat] ${name}: error on tick #${this._retries}: ${String(error)}`);
        const onError = this.config.onError;
        if (typeof onError === 'function') {
            onError(error);
        } else if (onError === 'skip') {
            void logger.warn(`[heartbeat] Skipping error: ${String(error)}`);
        } else if (this._retries >= this.config.maxRetries) { // "retry"
            void logger.error(
                `[heartbeat] ${name}: max retries (${this.config.maxRetries}) reached, skipping.`
            );
            this._retries = 0;
        }
    }

    /** Execute one heartbeat tick — run the agent and return its result text. */
    private async _tick(): Promise<string> {
        const prompt = this.config.prompt || 'Run your scheduled check.';
        const result = await this.agent.start(prompt);
        return result ? String(result) : '';
    }

    /** Deliver the result via callback or log. */
    private _deliver(result: string): void {
        if (this.config.onResult) {
            this.config.onResult(result);
        } else {
            void logger.info(`[heartbeat] ${this.agent.name ?? 'agent'}: ${result.slice(0, 200)}`);
        }
    }
}
